common/src/main/java/com/microsoft/identity/common/logging/Logger.java

```diff
@@ -0,0 +1,575 @@
+import java.util.Date;
+import java.util.Locale;
+import java.util.TimeZone;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+public class Logger {
+
+    private static final String CUSTOM_LOG_ERROR = "Custom log failed to log message:%s";
+    private static ExecutorService sLogExecutor = Executors.newSingleThreadExecutor();
+    private static final Logger INSTANCE = new Logger();
+    private static final String DATE_FORMAT = "yyyy-MM-dd HH:mm:ss";
+
+    // Turn on the VERBOSE level logging by default.
+    private LogLevel mLogLevel = LogLevel.VERBOSE;
+    private ILoggerCallback mExternalLogger;
+
+    // Disable to log PII by default.
+    private static boolean sAllowPii = false;
+    // Disable to Logcat logging by default.
+    private static boolean sAllowLogcat = false;
+
+    /**
+     * Enum class for LogLevel that the sdk recognizes.
+     */
+    public enum LogLevel {
+        /**
+         * Error level logging.
+         */
+        ERROR,
+        /**
+         * Warn level logging.
+         */
+        WARN,
+        /**
+         * Info level logging.
+         */
+        INFO,
+        /**
+         * Verbose level logging.
+         */
+        VERBOSE
+    }
+
+    /**
+     * @return The single instance of {@link Logger}.
+     */
+    public static Logger getInstance() {
+        return INSTANCE;
+    }
+
+    /**
+     * Enable/Disable log message with PII (personal identifiable information) info.
+     * By default, the SDK doesn't log any PII.
+     *
+     * @param allowPii True if enabling PII info to be logged, false otherwise.
+     */
+    public static void setAllowPii(final boolean allowPii) {
+        sAllowPii = allowPii;
+    }
+
+    /**
+     * Enable/Disable the Android logcat logging. By default, the sdk disables it.
+     *
+     * @param allowLogcat True if enabling the logcat logging, false otherwise.
+     */
+    public static void setAllowLogcat(final boolean allowLogcat) {
+        sAllowLogcat = allowLogcat;
+    }
+
+    /**
+     * Get if log PII is enabled.
+     *
+     * @return true if pii is allowed in logging, false otherwise.
+     */
+    public static boolean getAllowPii() {
+        return sAllowPii;
+    }
+
+    /**
+     * Get if logcat is enabled.
+     *
+     * @return true if logcat turned on, false otherwise.
+     */
+    public static boolean getAllowLogcat() {
+        return sAllowLogcat;
+    }
+
+    /**
+     * Set the log level for diagnostic purpose. By default, the sdk enables the verbose level
+     * logging.
+     *
+     * @param logLevel The {@link LogLevel} to be enabled for the diagnostic logging.
+     */
+    public void setLogLevel(final LogLevel logLevel) {
+        mLogLevel = logLevel;
+    }
+
+    /**
+     * Set the custom logger. Configures external logging to configure a callback that
+     * the sdk will use to pass each log message. Overriding the logger callback is not allowed.
```

Comment:
Do you mean overriding this method? Or resetting the external logger? Looking at the code it seems that both are actually allowed

Reply:
This is the exact Logger implementation from common, all I did was move the implementation

Reply:
From internal. There's no new code here

Reply:
I see - so the original documentation seems wrong then because code clearly allows overriding the method as well as resetting the logger. I think this is something that we need to think about and decide whether we actually want to allow resetting the logger. For apps that consume multiple of our libraries such as ad-accounts and MSAL and they still use the old Logger API from WPJ and MSAL to register their respective external logger, it seems that these loggers would now overwrite each other.....should we allow that?

Reply:
This of course assuming that we are delegating WPJ logger to common's logger

Reply:
This is an interesting question - how many applications is that?  I think it's 3 - Authenticator, CP, and BrokerHost.  But there may be more (teams, etc)... we should enumerate these problems.

Reply:
If ADAL's also going to use these then that would include basically any app that dual stacks ADAL and MSAL